File: sample-app-java/app/src/main/java/com/bluejeans/android/sdksample/menu/MenuFragment.java
/*
 * Copyright (c) 2021 Blue Jeans Network, Inc. All rights reserved.
 */
package com.bluejeans.android.sdksample.menu;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SwitchCompat;

import com.bjnclientcore.media.VideoSource;
import com.bluejeans.android.sdksample.R;
import com.bluejeans.android.sdksample.SampleApplication;
import com.bluejeans.bluejeanssdk.meeting.MeetingService;
import com.bluejeans.rxextensions.ObservableValueWithOptional;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.button.MaterialButton;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.Disposable;

public class MenuFragment extends BottomSheetDialogFragment {
    private static final String TAG = "MenuFragment";

    private IMenuCallback mIMenuCallback;
    private boolean mIsWaitingRoomEnabled;
    private MaterialButton mMbVideoLayout, mMbAudioDevice, mMbVideoDevice, mMbIscStreamStyle, mMbIscUseCases;
    private Button mViewWaitingRoom;
    private String mVideoLayout = "";
    private String mCurrentAudioDevice = "";
    private String mCurrentVideoDevice = "";
    private String mCurrentStreamStyle = "";
    private boolean mClosedCaptionState = false;
    private boolean mHDCaptureState = false;
    private boolean mHDReceiveState = false;
    private boolean mCustomVideoState = false;
    private boolean isIscSelected = false;
    private SwitchCompat mSwitchClosedCaption, mSwitchWaitingRoom, mSwitchHDCapture, mSwitch720Receive, mSwitchCustomVideo;
    private TextView mTvIscUseCase, mTvVideoStreamStyle;
    private LinearLayout mWaitingRoomLayout;

    private Disposable mWaitingRoomEnablementDisposable;

    private Disposable disposable;

    public interface IMenuCallback {
        void showVideoLayoutView(String videoLayoutName);

        void showAudioDeviceView();

        void showVideoDeviceView();

        void showIscStreamStyleView();

        void showVideoStreamUseCases();

        void handleClosedCaptionSwitchEvent(Boolean enabled);

        void handleHDCaptureSwitchEvent(Boolean isChecked);

        void handleHDReceiveSwitchEvent(Boolean isChecked);

        void showWaitingRoom();

        void setWaitingRoomEnabled(boolean enabled);

        void setCustomVideoSource(boolean isCustom);
    }

    public MenuFragment(IMenuCallback iMenuCallback, boolean isWaitingRoomEnabled, boolean isIscSelected) {
        mIMenuCallback = iMenuCallback;
        mIsWaitingRoomEnabled = isWaitingRoomEnabled;
        this.isIscSelected = isIscSelected;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_option_menu_dialog, container, false);
    }

    @Override
    public void onStart() {
        super.onStart();
        BottomSheetBehavior<View> bottomSheetBehavior = BottomSheetBehavior.from((View) requireView().getParent());
        bottomSheetBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews(view);

        disposable = SampleApplication.getBlueJeansSDK().getMeetingService()
                .getVideoLayout()
                .getRxObservable()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(videoLayoutOptional -> {
                    if (videoLayoutOptional.getValue() != null && videoLayoutOptional.getValue() instanceof MeetingService.VideoLayout.Custom) {
                        mTvIscUseCase.setVisibility(View.VISIBLE);
                        mMbIscUseCases.setVisibility(View.VISIBLE);
                        mMbIscStreamStyle.setVisibility(View.VISIBLE);
                        mTvVideoStreamStyle.setVisibility(View.VISIBLE);

                        mMbIscStreamStyle.setOnClickListener(view1 -> {
                            mIMenuCallback.showIscStreamStyleView();
                            dismiss();
                        });

                        mMbIscUseCases.setOnClickListener(view1 -> {
                            mIMenuCallback.showVideoStreamUseCases();
                            dismiss();
                        });
                    } else {
                        mTvIscUseCase.setVisibility(View.GONE);
                        mMbIscUseCases.setVisibility(View.GONE);
                        mMbIscStreamStyle.setVisibility(View.GONE);
                        mTvVideoStreamStyle.setVisibility(View.GONE);
                    }
                }, err -> {
                    Log.e(TAG, "Error: " + err.getLocalizedMessage());
                });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mWaitingRoomEnablementDisposable != null && !mWaitingRoomEnablementDisposable.isDisposed())
            mWaitingRoomEnablementDisposable.dispose();
    }

    public void updateVideoLayout(String videoLayout) {
        this.mVideoLayout = videoLayout;
        updateView();
    }

    public void updateAudioDevice(String currentAudioDevice) {
        this.mCurrentAudioDevice = currentAudioDevice;
        updateView();
    }

    public void updateVideoDevice(String currentVideoDevice) {
        this.mCurrentVideoDevice = currentVideoDevice;
        updateView();
    }

    public void updateVideoStreamStyle(String currentStreamStyle) {
        this.mCurrentStreamStyle = currentStreamStyle;
        updateView();
    }

    public void updateClosedCaptionSwitchState(boolean isClosedCaptionActive) {
        mClosedCaptionState = isClosedCaptionActive;
    }

    public void updateCustomVideoSwitchState(boolean isCustomVideo) {
        mCustomVideoState = isCustomVideo;
    }

    public void updateHDCaptureState(boolean captureHD) {
        mHDCaptureState = captureHD;
        mSwitchHDCapture.setChecked(mHDCaptureState);
    }

    private void initViews(View view) {
        mMbVideoLayout = view.findViewById(R.id.mbVideoLayout);
        mMbAudioDevice = view.findViewById(R.id.mbAudioDevice);
        mMbVideoDevice = view.findViewById(R.id.mbVideoDevice);
        mSwitchClosedCaption = view.findViewById(R.id.swClosedCaption);
        mSwitchHDCapture = view.findViewById(R.id.swHDCapture);
        mSwitch720Receive = view.findViewById(R.id.swHDReceive);
        mSwitchCustomVideo = view.findViewById(R.id.swCustomVideo);
        mMbIscStreamStyle = view.findViewById(R.id.mbVideoStreamStyle);
        mMbIscUseCases = view.findViewById(R.id.mbIscUseCases);
        mTvIscUseCase = view.findViewById(R.id.tvIscUseCases);
        mTvVideoStreamStyle = view.findViewById(R.id.tvStreamStyle);

        if (SampleApplication.getBlueJeansSDK().getBlueJeansClient().getMeetingSession() != null &&
                SampleApplication.getBlueJeansSDK().getBlueJeansClient().getMeetingSession().isModerator()) {
            mWaitingRoomLayout = view.findViewById(R.id.llWaitingRoom);
            mWaitingRoomLayout.setVisibility(View.VISIBLE);

            mViewWaitingRoom = view.findViewById(R.id.btnShowWaitingRoom);
            mSwitchWaitingRoom = view.findViewById(R.id.swWaitingRoom);

            boolean isWaitingRoomCapable = SampleApplication.getBlueJeansSDK().getMeetingService().getModeratorWaitingRoomService().isWaitingRoomCapable().getValue();
            if (isWaitingRoomCapable == true) {

                mViewWaitingRoom.setOnClickListener(view1 -> {
                    mIMenuCallback.showWaitingRoom();
                    dismiss();
                });

                mWaitingRoomEnablementDisposable = SampleApplication.getBlueJeansSDK().getMeetingService().getModeratorWaitingRoomService().isWaitingRoomEnabled()
                        .getRxObservable().observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                isEnabled -> {
                                    if (isEnabled.getValue()) {
                                        mSwitchWaitingRoom.setChecked(true);
                                    } else {
                                        mSwitchWaitingRoom.setChecked(false);
                                    }
                                }, err -> {
                                    Log.e(TAG, "Error occured while getting isWaitingRoomEnabled value");
                                }
                        );

                mSwitchWaitingRoom.setOnCheckedChangeListener((buttonView, isChecked) -> mIMenuCallback.setWaitingRoomEnabled(isChecked));
            } else if (isWaitingRoomCapable == false) {
                mViewWaitingRoom.setEnabled(false);
                mSwitchWaitingRoom.setEnabled(false);
            }
        }

        mMbVideoLayout.setOnClickListener(view1 -> {
            mIMenuCallback.showVideoLayoutView(mMbVideoLayout.getText().toString());
            dismiss();
        });

        mMbAudioDevice.setOnClickListener(view1 -> {
            mIMenuCallback.showAudioDeviceView();
            dismiss();
        });
        mMbVideoDevice.setOnClickListener(view1 -> {
            if (SampleApplication.getBlueJeansSDK().getCustomVideoSourceService().getCurrentVideoSource().getValue() == VideoSource.Custom.INSTANCE) {
                Toast.makeText(getContext(), getResources().getString(R.string.no_video_device_change), Toast.LENGTH_LONG)
                        .show();
            } else {
                mIMenuCallback.showVideoDeviceView();
                dismiss();
            }
        });

        if (this.mIsWaitingRoomEnabled) {
            mSwitchWaitingRoom.setChecked(this.mIsWaitingRoomEnabled);
        }

        if (isIscSelected) {
            mTvIscUseCase.setVisibility(View.VISIBLE);
            mMbIscUseCases.setVisibility(View.VISIBLE);
            mMbIscStreamStyle.setVisibility(View.VISIBLE);
            mTvVideoStreamStyle.setVisibility(View.VISIBLE);

            mMbIscStreamStyle.setOnClickListener(view1 -> {
                mIMenuCallback.showIscStreamStyleView();
                dismiss();
            });

            mMbIscUseCases.setOnClickListener(view1 -> {
                mIMenuCallback.showVideoStreamUseCases();
                dismiss();
            });
        } else {
            mTvIscUseCase.setVisibility(View.GONE);
            mMbIscUseCases.setVisibility(View.GONE);
            mMbIscStreamStyle.setVisibility(View.GONE);
            mTvVideoStreamStyle.setVisibility(View.GONE);
        }

        ObservableValueWithOptional<Boolean> closedCaptionFeatureObservable = SampleApplication.getBlueJeansSDK().getMeetingService()
                .getClosedCaptioningService().isClosedCaptioningAvailable();
        if (closedCaptionFeatureObservable.getValue() == Boolean.TRUE) {
            mSwitchClosedCaption.setChecked(mClosedCaptionState);
            mSwitchClosedCaption.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (buttonView.isPressed()) {
                    mIMenuCallback.handleClosedCaptionSwitchEvent(isChecked);
                    mClosedCaptionState = isChecked;
                    dismiss();
                }
            });
            mSwitchClosedCaption.setVisibility(View.VISIBLE);
        } else {
            mSwitchClosedCaption.setVisibility(View.GONE);
        }

        mHDCaptureState = SampleApplication.getBlueJeansSDK().getVideoDeviceService().is720pVideoCaptureEnabled().getValue();
        mSwitchHDCapture.setChecked(mHDCaptureState);
        mSwitchHDCapture.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (buttonView.isPressed()) {
                mIMenuCallback.handleHDCaptureSwitchEvent(isChecked);
            }
        });

        if (SampleApplication.getBlueJeansSDK().getVideoDeviceService().is720pVideoReceiveEnabled().getValue() != null) {
            mHDReceiveState = SampleApplication.getBlueJeansSDK().getVideoDeviceService().is720pVideoReceiveEnabled().getValue();
            mSwitch720Receive.setChecked(mHDReceiveState);
        }
        mSwitch720Receive.setOnCheckedChangeListener(((buttonView, isChecked) -> {
            if (buttonView.isPressed()) {
                mIMenuCallback.handleHDReceiveSwitchEvent(isChecked);
            }
        }));

        mSwitchCustomVideo.setChecked(mCustomVideoState);
        mSwitchCustomVideo.setOnCheckedChangeListener(((buttonView, isChecked) -> {
            mIMenuCallback.setCustomVideoSource(isChecked);
        }));
        updateView();
    }

    private void updateView() {
        if (mMbVideoLayout != null) {
            mMbVideoLayout.setText(mVideoLayout);
        }
        if (mMbAudioDevice != null) {
            mMbAudioDevice.setText(mCurrentAudioDevice);
        }
        if (mMbVideoDevice != null) {
            mMbVideoDevice.setText(mCurrentVideoDevice);
        }
        if (mMbIscStreamStyle != null) {
            mMbIscStreamStyle.setText(mCurrentStreamStyle);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mSwitchClosedCaption.setChecked(mClosedCaptionState);
        mSwitchHDCapture.setChecked(mHDCaptureState);
    }
}

File: sample-app-kotlin/app/src/main/java/com/bluejeans/android/sdksample/menu/MenuFragment.kt
/*
 * Copyright (c) 2021 Blue Jeans Network, Inc. All rights reserved.
 */
package com.bluejeans.android.sdksample.menu

import android.media.MediaRecorder.VideoSource
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintSet
import com.bjnclientcore.ui.util.extensions.gone
import com.bjnclientcore.ui.util.extensions.visible
import com.bluejeans.android.sdksample.R
import com.bluejeans.android.sdksample.SampleApplication
import com.bluejeans.android.sdksample.databinding.FragmentOptionMenuDialogBinding
import com.bluejeans.bluejeanssdk.meeting.MeetingService
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.kotlin.plusAssign
import timber.log.Timber

class MenuFragment(
    private val menuCallBack: IMenuCallback,
    private val isWaitingRoomEnabled: Boolean,
    private val isIscSelected: Boolean
) : BottomSheetDialogFragment() {

    private val TAG = "MenuFragment"

    private var videoLayout = ""
    private var currentAudioDevice = ""
    private var currentVideoDevice = ""
    private var currentStreamStyle = ""
    private var currentIscUseCase = ""
    private var closedCaptionState = false
    private var hdCaptureState = false
    private var isCustomVideo = false
    private var menuFragmentBinding: FragmentOptionMenuDialogBinding? = null
    private var bottomSheetBehavior: BottomSheetBehavior<View>? = null

    private val disposable: CompositeDisposable = CompositeDisposable()

    interface IMenuCallback {
        fun showVideoLayoutView(videoLayoutName: String)
        fun showAudioDeviceView()
        fun showVideoDeviceView()
        fun showIscStreamStyleView()
        fun showIscUseCases()
        fun handleClosedCaptionSwitchEvent(isChecked: Boolean)
        fun handleHDCaptureSwitchEvent(isChecked: Boolean)
        fun handleHDReceiveSwitchEvent(isChecked: Boolean)
        fun showWaitingRoom()
        fun setWaitingRoomEnabled(enabled: Boolean)
        fun setCustomVideoSource(isCustom: Boolean)
    }

    override fun onStart() {
        super.onStart()
        bottomSheetBehavior = BottomSheetBehavior.from(requireView().parent as View)
        bottomSheetBehavior?.state = BottomSheetBehavior.STATE_EXPANDED
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        menuFragmentBinding = FragmentOptionMenuDialogBinding.inflate(inflater, container, false)

        if (isWaitingRoomEnabled) {
            menuFragmentBinding!!.swWaitingRoom.isChecked = isWaitingRoomEnabled
        }

        disposable += SampleApplication.blueJeansSDK.meetingService.moderatorWaitingRoomService.isWaitingRoomEnabled
            .rxObservable.observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                {
                    it.value?.let { isChecked ->
                        menuFragmentBinding!!.swWaitingRoom.isChecked = isChecked
                    }
                },
                {
                    Timber.tag(TAG).e(it.message)
                })

        return menuFragmentBinding!!.root
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.clear()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()

        disposable += SampleApplication.blueJeansSDK.meetingService.videoLayout
            .rxObservable
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                it.value?.let { videoLayout ->
                    if (videoLayout == MeetingService.VideoLayout.Custom) {
                        menuFragmentBinding?.mbIscUseCases?.visibility = View.VISIBLE
                        menuFragmentBinding?.tvIscUseCases?.visibility = View.VISIBLE
                        menuFragmentBinding?.mbVideoStreamStyle?.visibility = View.VISIBLE
                        menuFragmentBinding?.tvStreamStyle?.visibility = View.VISIBLE
                        setUIConstraints()

                        menuFragmentBinding?.mbIscUseCases?.setOnClickListener {
                            menuCallBack.showIscUseCases()
                            dismiss()
                        }
                        menuFragmentBinding?.mbVideoStreamStyle?.setOnClickListener {
                            menuCallBack.showIscStreamStyleView()
                            dismiss()
                        }
                    } else {
                        menuFragmentBinding?.mbIscUseCases?.visibility = View.GONE
                        menuFragmentBinding?.tvIscUseCases?.visibility = View.GONE
                        menuFragmentBinding?.mbVideoStreamStyle?.visibility = View.GONE
                        menuFragmentBinding?.tvStreamStyle?.visibility = View.GONE
                    }
                }
            }, {
                Timber.tag(TAG).e("Error ${it.stackTraceToString()}")
            })

    }

    fun updateVideoLayout(videoLayout: String?) {
        if (videoLayout != null) {
            this.videoLayout = videoLayout
        }
        updateView()
    }

    fun updateAudioDevice(currentAudioDevice: String?) {
        this.currentAudioDevice = currentAudioDevice!!
        updateView()
    }

    fun updateVideoDevice(currentVideoDevice: String?) {
        this.currentVideoDevice = currentVideoDevice!!
        updateView()
    }

    fun updateVideoStreamStyle(currentStreamStyle: String) {
        this.currentStreamStyle = currentStreamStyle
        updateView()
    }

    fun updateIscUseCase(useCase: String) {
        this.currentIscUseCase = useCase;
        updateView()
    }

    fun updateClosedCaptionSwitchState(isClosedCaptionActive: Boolean) {
        closedCaptionState = isClosedCaptionActive
    }

    fun updateHDCaptureState(captureHD: Boolean) {
        hdCaptureState = captureHD
        menuFragmentBinding?.swHDCapture?.isChecked = hdCaptureState
    }

    fun updateCustomVideoSwitchState(isCustomVideo: Boolean) {
        this.isCustomVideo = isCustomVideo
    }

    private fun initViews() {
        menuFragmentBinding?.mbVideoLayout?.setOnClickListener {
            menuCallBack.showVideoLayoutView(menuFragmentBinding!!.mbVideoLayout.text as String)
            dismiss()
        }
        menuFragmentBinding?.mbAudioDevice?.setOnClickListener {
            menuCallBack.showAudioDeviceView()
            dismiss()
        }

        menuFragmentBinding?.mbVideoDevice?.setOnClickListener {
            if (SampleApplication.blueJeansSDK.customVideoSourceService.currentVideoSource.value == com.bjnclientcore.media.VideoSource.Custom) {
                Toast.makeText(context, resources.getString(R.string.no_video_device_change), Toast.LENGTH_LONG).show()
            } else {
                menuCallBack.showVideoDeviceView()
                dismiss()
            }
        }

        if (isIscSelected) {
            menuFragmentBinding?.mbIscUseCases?.visibility = View.VISIBLE
            menuFragmentBinding?.tvIscUseCases?.visibility = View.VISIBLE
            menuFragmentBinding?.mbVideoStreamStyle?.visibility = View.VISIBLE
            menuFragmentBinding?.tvStreamStyle?.visibility = View.VISIBLE
            setUIConstraints()

            menuFragmentBinding?.mbIscUseCases?.setOnClickListener {
                menuCallBack.showIscUseCases()
                dismiss()
            }

            menuFragmentBinding?.mbVideoStreamStyle?.setOnClickListener {
                menuCallBack.showIscStreamStyleView()
                dismiss()
            }
        } else {
            menuFragmentBinding?.mbIscUseCases?.visibility = View.GONE
            menuFragmentBinding?.tvIscUseCases?.visibility = View.GONE
            menuFragmentBinding?.mbVideoStreamStyle?.visibility = View.GONE
            menuFragmentBinding?.tvStreamStyle?.visibility = View.GONE
        }

        val closedCaptionFeatureObservable =
            SampleApplication.blueJeansSDK.meetingService.closedCaptioningService.isClosedCaptioningAvailable
        if (closedCaptionFeatureObservable.value == true) {
            menuFragmentBinding?.swClosedCaption?.isChecked = closedCaptionState
            menuFragmentBinding?.swClosedCaption?.setOnCheckedChangeListener { buttonView, isChecked ->
                if (buttonView.isPressed) {
                    menuCallBack.handleClosedCaptionSwitchEvent(isChecked)
                    closedCaptionState = isChecked
                    dismiss()
                }
            }
            menuFragmentBinding?.swClosedCaption?.visible()
        } else {
            menuFragmentBinding?.swClosedCaption?.gone()
        }


        hdCaptureState =
            SampleApplication.blueJeansSDK.videoDeviceService.is720pVideoCaptureEnabled.value
        menuFragmentBinding?.swHDCapture?.isChecked = hdCaptureState
        menuFragmentBinding?.swHDCapture?.setOnCheckedChangeListener { buttonView, isChecked ->
            if (buttonView.isPressed) {
                menuCallBack.handleHDCaptureSwitchEvent(isChecked)
            }
        }

        menuFragmentBinding?.swHDReceive?.isChecked =
            SampleApplication.blueJeansSDK.videoDeviceService.is720pVideoReceiveEnabled.value == true
        menuFragmentBinding?.swHDReceive?.setOnCheckedChangeListener { buttonView, isChecked ->
            menuCallBack.handleHDReceiveSwitchEvent(isChecked)
        }

        if (SampleApplication.blueJeansSDK.meetingService.moderatorControlsService.isModeratorControlsAvailable.value == true) {
            menuFragmentBinding?.llWaitingRoom?.visibility = View.VISIBLE

            if (SampleApplication.blueJeansSDK.meetingService.moderatorWaitingRoomService.isWaitingRoomCapable.value == true) {
                menuFragmentBinding?.btnShowWaitingRoom?.setOnClickListener {
                    menuCallBack.showWaitingRoom()
                    dismiss()
                }

                menuFragmentBinding?.swWaitingRoom?.setOnCheckedChangeListener { compoundButton, b ->
                    menuCallBack.setWaitingRoomEnabled(b)
                }
            } else if (SampleApplication.blueJeansSDK.meetingService.moderatorWaitingRoomService.isWaitingRoomCapable.value == false) {
                menuFragmentBinding?.btnShowWaitingRoom?.isEnabled = false
                menuFragmentBinding?.swWaitingRoom?.isEnabled = false
            }
        }

        menuFragmentBinding?.swCustomVideo?.isChecked = isCustomVideo
        menuFragmentBinding?.swCustomVideo?.setOnCheckedChangeListener { _, isChecked ->
            menuCallBack.setCustomVideoSource(isChecked)
        }

        updateView()
    }

    private fun updateView() {
        menuFragmentBinding?.let {
            it.mbVideoLayout.text = videoLayout
            it.mbAudioDevice.text = currentAudioDevice
            it.mbVideoDevice.text = currentVideoDevice
            it.mbVideoStreamStyle.text = currentStreamStyle
            it.mbIscUseCases.text = currentIscUseCase
        }
    }

    private fun setUIConstraints() {
        menuFragmentBinding?.let {
            var set = ConstraintSet()
            val btnApplyStreamStyle = it.mbVideoStreamStyle
            val horizontalLine = it.horizontalLine
            set.clone(it.controls)
            set.connect(
                btnApplyStreamStyle.id,
                ConstraintSet.BOTTOM,
                it.tvIscUseCases.id,
                ConstraintSet.TOP
            )

            set.applyTo(it.controls)

            set = ConstraintSet()
            set.clone(it.controls)
            set.connect(
                horizontalLine.id,
                ConstraintSet.TOP,
                it.mbIscUseCases.id,
                ConstraintSet.BOTTOM
            )
            set.applyTo(it.controls)
        }
    }

    override fun onResume() {
        super.onResume()
        menuFragmentBinding?.swClosedCaption?.isChecked = closedCaptionState
        menuFragmentBinding?.swHDCapture?.isChecked = hdCaptureState
    }
}